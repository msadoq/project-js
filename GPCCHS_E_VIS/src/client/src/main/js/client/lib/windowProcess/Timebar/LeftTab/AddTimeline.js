// ====================================================================
// HISTORY
// VERSION : 1.1.2 : DM : #5828 : 22/03/2017 : Reorganized files and folders in windowProcess/Timebar
// VERSION : 1.1.2 : DM : #5828 : 29/03/2017 : Replace sessionId by sessionName in timeline definition
// END-HISTORY
// ====================================================================

import React, { PureComponent } from 'react';
import { PropTypes } from 'prop-types';
import {
  Form,
  ButtonGroup,
  Button,
} from 'react-bootstrap';
import _get from 'lodash/get';
import _set from 'lodash/set';
import { reduxForm } from 'redux-form';
import TimelineFields from './TimelineFields';

class AddTimeline extends PureComponent {

  static propTypes = {
    handleSubmit: PropTypes.func.isRequired,
    timelines: PropTypes.arrayOf(
      PropTypes.shape({
        color: PropTypes.string,
        id: PropTypes.string.isRequired,
        kind: PropTypes.string.isRequired,
        uuid: PropTypes.string.isRequired,
        offset: PropTypes.number.isRequired,
        sessionName: PropTypes.string.isRequired,
      })
    ).isRequired,
    sessions: PropTypes.arrayOf(PropTypes.object).isRequired,
    pristine: PropTypes.bool.isRequired,
    submitting: PropTypes.bool.isRequired,
    valid: PropTypes.bool.isRequired,
  };

  state = {
    disableSubmit: false,
  }

  disableSubmit = (flag) => {
    this.setState({
      disableSubmit: flag,
    });
  }

  render() {
    const {
      sessions,
      timelines,
      pristine,
      submitting,
      valid,
      handleSubmit,
    } = this.props;
    const { disableSubmit } = this.state;

    return (
      <Form horizontal onSubmit={handleSubmit}>
        <TimelineFields
          sessions={sessions}
          timelines={timelines}
          disableSubmit={this.disableSubmit}
        />
        <div className="text-right">
          <ButtonGroup>
            <Button
              bsStyle="success"
              type="submit"
              disabled={pristine || submitting || !valid || disableSubmit}
            >
              Submit
            </Button>
          </ButtonGroup>
        </div>
      </Form>
    );
  }
}

const requiredFields = ['id', 'sessionName'];
const validate = (values = {}) => {
  const errors = {};
  requiredFields.forEach((fieldPath) => {
    if (!_get(values, fieldPath)) {
      _set(errors, fieldPath, 'Required');
    }
  });

  return errors;
};

export default reduxForm({
  validate,
  enableReinitialize: true,
})(AddTimeline);
