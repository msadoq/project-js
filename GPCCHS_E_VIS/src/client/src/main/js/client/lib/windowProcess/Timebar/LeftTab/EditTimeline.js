// import moment from 'moment';
import React, { PureComponent, PropTypes } from 'react';
import _get from 'lodash/get';
import _set from 'lodash/set';
import {
  Form,
  ButtonGroup,
  Button,
} from 'react-bootstrap';
import { reduxForm } from 'redux-form';
import TimelineFields from './TimelineFields';

class EditTimeline extends PureComponent {

  static propTypes = {
    handleSubmit: PropTypes.func.isRequired,
    timelines: PropTypes.arrayOf(
      PropTypes.shape({
        color: PropTypes.string,
        id: PropTypes.string.isRequired,
        kind: PropTypes.string.isRequired,
        uuid: PropTypes.string.isRequired,
        offset: PropTypes.number.isRequired,
        sessionName: PropTypes.string.isRequired,
      })
    ).isRequired,
    sessions: PropTypes.arrayOf(
      PropTypes.shape({
        name: PropTypes.string.isRequired,
        delta: PropTypes.number.isRequired,
        id: PropTypes.number.isRequired,
        missionEpoch: PropTypes.number.isRequired,
        timestamp: PropTypes.shape({
          ms: PropTypes.number,
          ps: PropTypes.number,
        }),
      })
    ).isRequired,
    masterId: PropTypes.string,
    id: PropTypes.string.isRequired,
    uuid: PropTypes.string.isRequired,
    pristine: PropTypes.bool.isRequired,
    submitting: PropTypes.bool.isRequired,
    valid: PropTypes.bool.isRequired,
  }

  static defaultProps = {
    masterId: null,
  }

  state = {
    disableSubmit: false,
  }

  disableSubmit = (flag) => {
    this.setState({
      disableSubmit: flag,
    });
  }

  render() {
    const {
      sessions,
      timelines,
      pristine,
      submitting,
      valid,
      handleSubmit,
      masterId,
      id,
      uuid,
    } = this.props;
    const { disableSubmit } = this.state;

    return (
      <Form horizontal onSubmit={handleSubmit}>

        <TimelineFields
          sessions={sessions}
          timelines={timelines}
          masterId={masterId}
          uuid={uuid}
          id={id}
          disableSubmit={this.disableSubmit}
        />
        <div className="text-right">
          <ButtonGroup>
            <Button
              bsStyle="success"
              type="submit"
              disabled={pristine || submitting || !valid || disableSubmit}
            >
              Submit
            </Button>
          </ButtonGroup>
        </div>
      </Form>
    );
  }
}

const requiredFields = ['id', 'sessionName'];
const validate = (values = {}) => {
  const errors = {};
  requiredFields.forEach((fieldPath) => {
    if (!_get(values, fieldPath)) {
      _set(errors, fieldPath, 'Required');
    }
  });

  return errors;
};

export default reduxForm({
  validate,
  enableReinitialize: true,
})(EditTimeline);
