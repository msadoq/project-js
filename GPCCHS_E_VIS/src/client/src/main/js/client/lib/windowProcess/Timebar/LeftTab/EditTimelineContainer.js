// ====================================================================
// HISTORY
// VERSION : 1.1.2 : DM : #5828 : 12/04/2017 : New GenericModal component displayed or not displayed at root (Window.js) AddTimeline and EditTimeline forms displayed through it.
// END-HISTORY
// ====================================================================

import { connect } from 'react-redux';
import { getSessions } from '../../../store/reducers/sessions';
import {
  updateId,
  updateOffset,
  updateSessionName,
  updateColor,
} from '../../../store/actions/timelines';
import {
  updateMasterId,
} from '../../../store/actions/timebars';
import {
  getTimebarTimelinesSelector,
} from '../../../store/selectors/timebars';
import {
  getTimebar,
} from '../../../store/reducers/timebars';
import {
  getTimeline,
} from '../../../store/reducers/timelines';
import EditTimelineWrapper from './EditTimelineWrapper';

const mapStateToProps = (state, { timebarUuid, timelineUuid }) => ({
  timelines: getTimebarTimelinesSelector(state, { timebarUuid }),
  sessions: getSessions(state),
  timebar: getTimebar(state, { timebarUuid }),
  timeline: getTimeline(state, { timelineUuid }),
});

const mapDispatchToProps = {
  updateId,
  updateOffset,
  updateSessionName,
  updateColor,
  updateMasterId,
};

export default connect(mapStateToProps, mapDispatchToProps)(EditTimelineWrapper);
