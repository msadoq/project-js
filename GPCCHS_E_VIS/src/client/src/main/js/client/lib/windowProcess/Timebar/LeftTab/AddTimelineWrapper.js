// ====================================================================
// HISTORY
// VERSION : 1.1.2 : DM : #5828 : 12/04/2017 : New GenericModal component displayed or not
//  displayed at root (Window.js) AddTimeline and EditTimeline forms displayed through it.
// VERSION : 1.1.2 : DM : #5828 : 13/04/2017 : Lint errors fixes and useless props removed in
//  LeftTab / AddTimelineWrapper.
// END-HISTORY
// ====================================================================

import React, { PureComponent, PropTypes } from 'react';
import { schemeCategory20b } from 'd3-scale';
import AddTimeline from './AddTimeline';

export default class AddTimelineWrapper extends PureComponent {
  static propTypes = {
    closeModal: PropTypes.func.isRequired,
    createNewTimeline: PropTypes.func.isRequired,
    updateMasterId: PropTypes.func.isRequired,
    updateOffset: PropTypes.func.isRequired,
    timebarUuid: PropTypes.string.isRequired,
    timelines: PropTypes.arrayOf(
      PropTypes.shape({
        color: PropTypes.string,
        id: PropTypes.string.isRequired,
        kind: PropTypes.string.isRequired,
        uuid: PropTypes.string.isRequired,
        offset: PropTypes.number.isRequired,
        sessionName: PropTypes.string.isRequired,
      })
    ).isRequired,
    sessions: PropTypes.arrayOf(
      PropTypes.shape({
        name: PropTypes.string.isRequired,
        delta: PropTypes.number.isRequired,
        id: PropTypes.number.isRequired,
        missionEpoch: PropTypes.number.isRequired,
        timestamp: PropTypes.shape({
          ms: PropTypes.number,
          ps: PropTypes.number,
        }),
      })
    ).isRequired,
  }

  willAddTimeline = (values) => {
    const {
      timebarUuid,
      timelines,
      updateMasterId,
      createNewTimeline,
      updateOffset,
      closeModal,
    } = this.props;

    const timelinesBeforeAdd = [].concat(timelines);

    createNewTimeline(
      timebarUuid,
      {
        kind: values.kind,
        id: values.id,
        sessionName: values.sessionName,
        color: values.color,
        offset: values.master ? 0 : parseInt(values.offset, 10),
      }
    );
    closeModal();
    if (values.master) {
      timelinesBeforeAdd.forEach(t =>
        updateOffset(t.uuid, t.offset - values.offset)
      );
      updateMasterId(timebarUuid, values.id);
    }
  }

  render() {
    const {
      timelines,
      sessions,
      timebarUuid,
    } = this.props;

    return (
      <AddTimeline
        form={`timebar-${timebarUuid}-addTrack`}
        sessions={sessions}
        timelines={timelines}
        onSubmit={this.willAddTimeline}
        // eslint-disable-next-line react-perf/jsx-no-new-object-as-prop, "DV6 TBC_CNES ReduxForm"
        initialValues={{
          id: '',
          color: schemeCategory20b[timelines.length % 20],
          kind: 'session',
          sessionName: sessions[0] ? sessions[0].name : '',
          offset: 0,
          master: false,
        }}
      />
    );
  }
}
