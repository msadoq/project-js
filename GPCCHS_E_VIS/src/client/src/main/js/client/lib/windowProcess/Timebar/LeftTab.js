import classnames from 'classnames';
import React, { PureComponent, PropTypes } from 'react';
import { Col, Glyphicon, Button } from 'react-bootstrap';
import { schemeCategory20b } from 'd3-scale';
import Timeline from './Timeline';
import Modal from '../common/Modal';
import EditTimeline from './timeline/EditTimeline';
import AddTimeline from './timeline/AddTimeline';
import styles from './Lefttab.css';
import { main } from '../ipc';

export default class LeftTab extends PureComponent {
  static propTypes = {
    collapseTimebar: PropTypes.func.isRequired,
    addAndMountTimeline: PropTypes.func.isRequired,
    unmountTimeline: PropTypes.func.isRequired,
    onTimelinesVerticalScroll: PropTypes.func.isRequired,
    updateId: PropTypes.func.isRequired,
    updateColor: PropTypes.func.isRequired,
    updateMasterId: PropTypes.func.isRequired,
    updateOffset: PropTypes.func.isRequired,
    updateSessionId: PropTypes.func.isRequired,
    updateTimebarId: PropTypes.func.isRequired,
    timelines: PropTypes.array.isRequired,
    sessions: PropTypes.array.isRequired,
    timebarUuid: PropTypes.string.isRequired,
    timebarName: PropTypes.string.isRequired,
    focusedPageId: PropTypes.string.isRequired,
    masterId: PropTypes.string,
    verticalScroll: PropTypes.number.isRequired,
  }

  constructor(...args) {
    super(...args);
    this.state = {
      willAdd: false,
      willEdit: false,
    };
  }

  componentDidUpdate() {
    this.timelinesEl.scrollTop = this.props.verticalScroll;
  }

  onWheel = (e) => {
    this.props.onTimelinesVerticalScroll(e, e.currentTarget);
  }

  willUnmountTimeline = (timebarUuid, timelineId) => {
    this.props.unmountTimeline(timebarUuid, timelineId);
  }

  toggleAddTimeline = (e) => {
    if (e) {
      e.preventDefault();
    }

    if (!this.state.willAdd) {
      // TODO send reload sessions request to main and on callback
      // refresh session list from DC before displaying form
      main.reloadSessions(() => {
        this.setState({ willAdd: true });
      });
    } else {
      this.setState({ willAdd: false });
    }
  }

  willAddTimeline = (values) => {
    const {
      timebarUuid,
      timelines,
      updateMasterId,
      addAndMountTimeline,
      updateOffset,
    } = this.props;

    const timelinesBeforeAdd = [].concat(timelines);

    addAndMountTimeline(
      timebarUuid,
      {
        kind: values.kind,
        id: values.id,
        sessionId: parseInt(values.sessionId, 10),
        color: values.color,
        offset: values.master ? 0 : parseInt(values.offset, 10),
      }
    );
    this.setState({
      willAdd: false,
    });

    if (values.master) {
      timelinesBeforeAdd.forEach(t =>
        updateOffset(t.timelineId, t.offset - values.offset)
      );
      updateMasterId(timebarUuid, values.id);
    }
  }

  hideModals = () => {
    this.setState({
      willAdd: false,
      willEdit: false,
    });
  }

  willEditTimeline = (timelineId, id) => {
    this.setState({
      willAdd: false,
      willEdit: true,
      editingId: id,
    });
  }

  editTimeline = (values) => {
    const {
      updateOffset,
      updateId,
      updateColor,
      timebarUuid,
      updateMasterId,
      masterId,
      timelines,
      updateSessionId,
    } = this.props;

    const timeline = timelines.find(x => x.timelineId === values.timelineId);
    const offset = parseInt(values.offset, 10);

    if (timeline.id !== values.id) {
      updateId(values.timelineId, values.id);
    }
    if (timeline.color !== values.color) {
      updateColor(values.timelineId, values.color);
    }
    if (timeline.sessionId !== parseInt(values.sessionId, 10)) {
      updateSessionId(values.timelineId, parseInt(values.sessionId, 10));
    }

    if (values.master && masterId !== values.id) {
      updateMasterId(timebarUuid, values.id);
      timelines.forEach((t) => {
        if (t.timelineId === values.timelineId) {
          return;
        }
        updateOffset(t.timelineId, t.offset - offset);
      });
      updateOffset(values.timelineId, 0);
    } else if (timeline.offset !== offset) {
      if ((masterId === values.id) || (values.master && masterId !== values.id)) {
        timelines.forEach((t) => {
          if (t.timelineId === values.timelineId) {
            return;
          }
          updateOffset(t.timelineId, t.offset - offset);
        });
        updateOffset(values.timelineId, 0);
      } else {
        updateOffset(values.timelineId, offset);
      }
    }
    this.setState({ editingId: null });
  }

  detach = (e) => {
    e.preventDefault();
    this.props.updateTimebarId(this.props.focusedPageId, null);
  }

  collapse = (e) => {
    e.preventDefault();
    this.props.collapseTimebar(this.props.focusedPageId, true);
  }

  render() {
    const {
      timelines,
      masterId,
      sessions,
      timebarName,
      timebarUuid,
    } = this.props;
    const {
      willAdd,
      willEdit,
      editingId,
    } = this.state;

    let noTrack;
    if (timelines.length === 0) {
      noTrack = <h5 className="text-center"><br /><b>No track to display</b></h5>;
    }
    const currentlyEditingTimeline = timelines.find(x => x.id === editingId);

    const editTrack = (
      <Modal
        title="Edit timeline"
        isOpened={currentlyEditingTimeline && willEdit}
        onClose={this.hideModals}
      >
        {currentlyEditingTimeline && <EditTimeline
          form={`timebar-${timebarUuid}-editTrack`}
          onSubmit={this.editTimeline}
          sessions={sessions}
          timelines={timelines}
          masterId={masterId}
          id={currentlyEditingTimeline.id}
          timelineId={currentlyEditingTimeline.timelineId}
          // eslint-disable-next-line react-perf/jsx-no-new-object-as-prop
          initialValues={{
            master: masterId === currentlyEditingTimeline.id,
            id: currentlyEditingTimeline.id,
            color: currentlyEditingTimeline.color,
            kind: currentlyEditingTimeline.kind,
            sessionId: typeof currentlyEditingTimeline.sessionId === 'number' ?
              currentlyEditingTimeline.sessionId.toString() : '',
            timelineId: currentlyEditingTimeline.timelineId,
            offset: currentlyEditingTimeline.offset,
          }}
        />}
      </Modal>
    );

    const addTrack = (
      <Modal
        title="Add timeline"
        isOpened={willAdd}
        onClose={this.toggleAddTimeline}
      >
        <AddTimeline
          form={`timebar-${timebarUuid}-addTrack`}
          sessions={sessions}
          timelines={timelines}
          onSubmit={this.willAddTimeline}
          // eslint-disable-next-line react-perf/jsx-no-new-object-as-prop
          initialValues={{
            id: '',
            color: schemeCategory20b[timelines.length % 20],
            kind: 'session',
            sessionId: typeof sessions[0].id === 'number' ?
              sessions[0].id.toString() : '',
            offset: 0,
            master: false,
          }}
        />
      </Modal>
    );

    return (
      <Col
        xs={3}
        className={classnames(
          styles.leftTab,
          'subdiv'
        )}
      >
        <div className={styles.leftTabTopPanel}>
          {editTrack}
          {addTrack}
          <Button
            title="Detach timebar"
            bsSize="xs"
            onClick={this.detach}
            className={classnames('btn-transp', styles.btnClose)}
          >
            <Glyphicon
              glyph="remove"
            />
          </Button>
          <Button
            title="Collapse timebar"
            bsSize="xs"
            onClick={this.collapse}
            className={classnames('btn-transp', styles.btnCollapse)}
          >
            <Glyphicon
              glyph="minus"
            />
          </Button>
          <h5 className={styles.timebarName}>
            <b>{timebarName}</b>
          </h5>
          <Button
            title="Add timeline"
            bsSize="xs"
            onClick={this.toggleAddTimeline}
            className={classnames('btn-transp', styles.addTimelineButton)}
          >
            <Glyphicon
              glyph="plus"
            />
          </Button>
        </div>
        {noTrack}
        <ul
          ref={(el) => { this.timelinesEl = el; }}
          className={styles.timelineUl}
          onScroll={this.props.onTimelinesVerticalScroll}
          onWheel={this.onWheel}
        >
          { timelines.map(v =>
            <Timeline
              key={v.i}
              offset={v.offset}
              timelinesLength={timelines.length}
              timebarUuid={timebarUuid}
              id={v.id}
              timelineId={v.timelineId}
              color={v.color}
              masterId={masterId}
              willEditTimeline={this.willEditTimeline}
              unmountTimeline={this.willUnmountTimeline}
            />
          )}
        </ul>
      </Col>
    );
  }
}
