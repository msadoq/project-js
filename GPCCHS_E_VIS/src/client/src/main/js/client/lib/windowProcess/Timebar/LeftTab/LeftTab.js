import classnames from 'classnames';
import React, { PureComponent, PropTypes } from 'react';
import { Col, Glyphicon, Button } from 'react-bootstrap';
import { schemeCategory20b } from 'd3-scale';
import Timeline from './Timeline';
import Modal from '../../common/Modal';
import EditTimeline from './EditTimeline';
import AddTimeline from './AddTimeline';
import styles from './LeftTab.css';
import { main } from '../../ipc';


export default class LeftTab extends PureComponent {
  static propTypes = {
    collapseTimebar: PropTypes.func.isRequired,
    createNewTimeline: PropTypes.func.isRequired,
    removeTimeline: PropTypes.func.isRequired,
    onTimelinesVerticalScroll: PropTypes.func.isRequired,
    updateId: PropTypes.func.isRequired,
    updateColor: PropTypes.func.isRequired,
    updateMasterId: PropTypes.func.isRequired,
    updateOffset: PropTypes.func.isRequired,
    updateSessionName: PropTypes.func.isRequired,
    updateTimebarId: PropTypes.func.isRequired,
    timelines: PropTypes.arrayOf(
      PropTypes.shape({
        color: PropTypes.string,
        id: PropTypes.string.isRequired,
        kind: PropTypes.string.isRequired,
        uuid: PropTypes.string.isRequired,
        offset: PropTypes.number.isRequired,
        sessionName: PropTypes.string.isRequired,
      })
    ).isRequired,
    sessions: PropTypes.arrayOf(
      PropTypes.shape({
        name: PropTypes.string.isRequired,
        delta: PropTypes.number.isRequired,
        id: PropTypes.number.isRequired,
        missionEpoch: PropTypes.number.isRequired,
        timestamp: PropTypes.shape({
          ms: PropTypes.number,
          ps: PropTypes.number,
        }),
      })
    ).isRequired,
    timebarUuid: PropTypes.string.isRequired,
    timebarName: PropTypes.string.isRequired,
    pageId: PropTypes.string.isRequired,
    masterId: PropTypes.string,
    verticalScroll: PropTypes.number.isRequired,
  }

  static defaultProps = {
    masterId: null,
  }

  constructor(...args) {
    super(...args);
    this.state = {
      willAdd: false,
      willEdit: false,
    };
  }

  componentDidUpdate() {
    this.timelinesEl.scrollTop = this.props.verticalScroll;
  }

  onWheel = (e) => {
    this.props.onTimelinesVerticalScroll(e, e.currentTarget);
  }

  willRemoveTimeline = (timebarUuid, timelineUuid) => {
    this.props.removeTimeline(timebarUuid, timelineUuid);
  }

  toggleAddTimeline = (e) => {
    if (e) {
      e.preventDefault();
    }

    if (!this.state.willAdd) {
      // TODO send reload sessions request to main and on callback
      // refresh session list from DC before displaying form
      main.reloadSessions(() => {
        this.setState({ willAdd: true });
      });
    } else {
      this.setState({ willAdd: false });
    }
  }

  willAddTimeline = (values) => {
    const {
      timebarUuid,
      timelines,
      updateMasterId,
      createNewTimeline,
      updateOffset,
    } = this.props;

    const timelinesBeforeAdd = [].concat(timelines);

    createNewTimeline(
      timebarUuid,
      {
        kind: values.kind,
        id: values.id,
        // sessionId: parseInt(values.sessionId, 10),
        sessionName: values.sessionName,
        color: values.color,
        offset: values.master ? 0 : parseInt(values.offset, 10),
      }
    );
    this.hideModals();

    if (values.master) {
      timelinesBeforeAdd.forEach(t =>
        updateOffset(t.uuid, t.offset - values.offset)
      );
      updateMasterId(timebarUuid, values.id);
    }
  }

  hideModals = () => {
    this.setState({
      willAdd: false,
      willEdit: false,
      editingId: null,
    });
  }

  willEditTimeline = (id) => {
    this.setState({
      willAdd: false,
      willEdit: true,
      editingId: id,
    });
  }

  editTimeline = (values) => {
    const {
      updateOffset,
      updateId,
      updateColor,
      timebarUuid,
      updateMasterId,
      masterId,
      timelines,
      updateSessionName,
    } = this.props;

    const timeline = timelines.find(x => x.uuid === values.uuid);
    const offset = parseInt(values.offset, 10);

    if (timeline.id !== values.id) {
      updateId(values.uuid, values.id);
    }
    if (timeline.color !== values.color) {
      updateColor(values.uuid, values.color);
    }
    // if (timeline.sessionId !== parseInt(values.sessionId, 10)) {
    if (timeline.sessionName !== values.sessionName) {
      updateSessionName(values.uuid, values.sessionName);
    }

    if (values.master && masterId !== values.id) {
      updateMasterId(timebarUuid, values.id);
      timelines.forEach((t) => {
        if (t.uuid === values.uuid) {
          return;
        }
        updateOffset(t.uuid, t.offset - offset);
      });
      updateOffset(values.uuid, 0);
    } else if (timeline.offset !== offset) {
      if ((masterId === values.id) || (values.master && masterId !== values.id)) {
        timelines.forEach((t) => {
          if (t.uuid === values.uuid) {
            return;
          }
          updateOffset(t.uuid, t.offset - offset);
        });
        updateOffset(values.uuid, 0);
      } else {
        updateOffset(values.uuid, offset);
      }
    }
    this.hideModals();
  }

  detach = (e) => {
    e.preventDefault();
    this.props.updateTimebarId(this.props.pageId, null);
  }

  collapse = (e) => {
    e.preventDefault();
    this.props.collapseTimebar(this.props.pageId, true);
  }

  render() {
    const {
      timelines,
      masterId,
      sessions,
      timebarName,
      timebarUuid,
    } = this.props;
    const {
      willAdd,
      willEdit,
      editingId,
    } = this.state;

    let noTrack;
    if (timelines.length === 0) {
      noTrack = <h5 className="text-center"><br /><b>No track to display</b></h5>;
    }
    const currentlyEditingTimeline = timelines.find(x => x.id === editingId);

    const editTrack = (
      <Modal
        title="Edit timeline"
        isOpened={currentlyEditingTimeline && willEdit}
        onClose={this.hideModals}
      >
        {currentlyEditingTimeline && <EditTimeline
          form={`timebar-${timebarUuid}-editTrack`}
          onSubmit={this.editTimeline}
          sessions={sessions}
          timelines={timelines}
          masterId={masterId}
          id={currentlyEditingTimeline.id}
          uuid={currentlyEditingTimeline.uuid}
          // eslint-disable-next-line react-perf/jsx-no-new-object-as-prop
          initialValues={{
            master: masterId === currentlyEditingTimeline.id,
            id: currentlyEditingTimeline.id,
            color: currentlyEditingTimeline.color,
            kind: currentlyEditingTimeline.kind,
            sessionName: currentlyEditingTimeline.sessionName,
            uuid: currentlyEditingTimeline.uuid,
            offset: currentlyEditingTimeline.offset,
          }}
        />}
      </Modal>
    );

    const addTrack = (
      <Modal
        title="Add timeline"
        isOpened={willAdd}
        onClose={this.toggleAddTimeline}
      >
        <AddTimeline
          form={`timebar-${timebarUuid}-addTrack`}
          sessions={sessions}
          timelines={timelines}
          onSubmit={this.willAddTimeline}
          // eslint-disable-next-line react-perf/jsx-no-new-object-as-prop
          initialValues={{
            id: '',
            color: schemeCategory20b[timelines.length % 20],
            kind: 'session',
            sessionId: typeof sessions[0].id === 'number' ?
              sessions[0].id.toString() : '',
            offset: 0,
            master: false,
          }}
        />
      </Modal>
    );

    return (
      <Col
        xs={3}
        className={classnames(
          styles.leftTab,
          'subdiv'
        )}
      >
        <div className={styles.leftTabTopPanel}>
          {editTrack}
          {addTrack}
          <Button
            title="Detach timebar"
            bsSize="xs"
            onClick={this.detach}
            className={classnames('btn-transp', styles.btnClose)}
          >
            <Glyphicon
              glyph="remove"
            />
          </Button>
          <Button
            title="Collapse timebar"
            bsSize="xs"
            onClick={this.collapse}
            className={classnames('btn-transp', styles.btnCollapse)}
          >
            <Glyphicon
              glyph="resize-small"
            />
          </Button>
          <h5 className={styles.timebarName}>
            <b>{timebarName}</b>
          </h5>
          <Button
            title="Add timeline"
            bsSize="xs"
            onClick={this.toggleAddTimeline}
            className={classnames('btn-transp', styles.addTimelineButton)}
          >
            <Glyphicon
              glyph="plus"
            />
          </Button>
        </div>
        {noTrack}
        <ul
          ref={(el) => { this.timelinesEl = el; }}
          className={styles.timelineUl}
          onScroll={this.props.onTimelinesVerticalScroll}
          onWheel={this.onWheel}
        >
          { timelines && timelines.map((v) => {
            const session = Object.values(sessions).find(s => s.name === v.sessionName);
            return (
              <Timeline
                key={v.id}
                offset={v.offset}
                timelinesLength={timelines.length}
                timebarUuid={timebarUuid}
                id={v.id}
                uuid={v.uuid}
                color={v.color}
                masterId={masterId}
                willEditTimeline={this.willEditTimeline}
                removeTimeline={this.willRemoveTimeline}
                sessionName={session ? session.name : 'no session'}
              />
            );
          }
          )}
        </ul>
      </Col>
    );
  }
}
