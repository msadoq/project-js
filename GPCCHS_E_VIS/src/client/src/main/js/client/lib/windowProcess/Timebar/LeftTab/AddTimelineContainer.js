// ====================================================================
// HISTORY
// VERSION : 1.1.2 : DM : #5828 : 12/04/2017 : New GenericModal component displayed or not displayed at root (Window.js) AddTimeline and EditTimeline forms displayed through it.
// END-HISTORY
// ====================================================================

import { connect } from 'react-redux';
import { getSessions } from '../../../store/reducers/sessions';
import {
  createNewTimeline,
  updateOffset,
} from '../../../store/actions/timelines';
import {
  updateMasterId,
} from '../../../store/actions/timebars';
import {
  getTimebarTimelinesSelector,
} from '../../../store/selectors/timebars';
import {
  getTimebar,
} from '../../../store/reducers/timebars';
import AddTimelineWrapper from './AddTimelineWrapper';

const mapStateToProps = (state, { timebarUuid }) => {
  const timebar = getTimebar(state, { timebarUuid });
  return {
    timelines: getTimebarTimelinesSelector(state, { timebarUuid }),
    sessions: getSessions(state),
    masterId: timebar.masterId,
  };
};

const mapDispatchToProps = {
  createNewTimeline,
  updateMasterId,
  updateOffset,
};

export default connect(mapStateToProps, mapDispatchToProps)(AddTimelineWrapper);
