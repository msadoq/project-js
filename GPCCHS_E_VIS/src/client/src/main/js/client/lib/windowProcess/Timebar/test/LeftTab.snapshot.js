import React from 'react';
import renderer from 'react-test-renderer';
import TimebarFixture from '../../../../test/__mocks__/Timebar';
import TimelinesFixture from '../../../../test/__mocks__/Timelines';
import SessionsFixture from '../../../../test/__mocks__/Sessions';
import LeftTab from '../LeftTab/LeftTab';

const propsStub = {
  addAndMountTimeline: () => null,
  unmountTimeline: () => null,
  onTimelinesVerticalScroll: () => null,
  updateId: () => null,
  updateMasterId: () => null,
  updateOffset: () => null,
  createNewTimeline: () => null,
  removeTimeline: () => null,
  updateTimebarId: () => null,
  updateSessionId: () => null,
  updateColor: () => null,
  collapseTimebar: () => null,
  timelines: TimelinesFixture,
  sessions: SessionsFixture,
  timebarUuid: TimebarFixture.timebarUuid,
  timebarName: TimebarFixture.id,
  focusedPageId: 'rezter-456789',
  masterId: TimebarFixture.masterId,
  verticalScroll: 0,
};

test('Lefttab renders correctly', () => {
  const tree = renderer.create(
    <LeftTab
      {...propsStub}
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
