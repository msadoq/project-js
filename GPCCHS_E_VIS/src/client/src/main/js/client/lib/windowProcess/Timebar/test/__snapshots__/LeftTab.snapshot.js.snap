exports[`test Lefttab renders correctly 1`] = `
<div
  className="leftTab subdiv col-xs-3">
  <div
    className="leftTabTopPanel">
    <div
      className="modal-container" />
    <div
      className="modal-container" />
    <button
      className="btn-transp btnClose btn btn-xs btn-default"
      disabled={false}
      onClick={[Function]}
      title="Detach timebar"
      type="button">
      <span
        className="glyphicon glyphicon-remove" />
    </button>
    <button
      className="btn-transp btnCollapse btn btn-xs btn-default"
      disabled={false}
      onClick={[Function]}
      title="Collapse timebar"
      type="button">
      <span
        className="glyphicon glyphicon-minus" />
    </button>
    <h5
      className="timebarName">
      <b>
        Timebar 01
      </b>
    </h5>
    <button
      className="btn-transp addTimelineButton btn btn-xs btn-default"
      disabled={false}
      onClick={[Function]}
      title="Add timeline"
      type="button">
      <span
        className="glyphicon glyphicon-plus" />
    </button>
  </div>
  <ul
    className="timelineUl"
    onScroll={[Function]}
    onWheel={[Function]}>
    <li
      className="timeline">
      <span
        className="master"
        title="Master timeline">
        M
      </span>
      <span
        style={
          Object {
            "paddingLeft": "20px",
          }
        }>
        Session 1
         | 
        Master
      </span>
      <span
        className="offset"
        title="Formatted offset HH:MM:ss:SSS">
        + 00:00:00.200
      </span>
      <span
        className="square"
        style={
          Object {
            "background": "#fa1289",
            "right": "26px",
          }
        } />
      <button
        className="editButton"
        onClick={[Function]}
        style={
          Object {
            "right": "0px",
          }
        }
        title="Upate this track">
        <span
          className="glyphicon glyphicon-edit" />
      </button>
    </li>
    <li
      className="timeline">
      <span
        style={
          Object {
            "paddingLeft": "0px",
          }
        }>
        Session 2
         | 
        Master
      </span>
      <span
        className="offset"
        title="Formatted offset HH:MM:ss:SSS">
        + 00:00:01.000
      </span>
      <span
        className="square"
        style={
          Object {
            "background": "#31b0d5",
            "right": "52px",
          }
        } />
      <button
        className="editButton"
        onClick={[Function]}
        style={
          Object {
            "right": "26px",
          }
        }
        title="Upate this track">
        <span
          className="glyphicon glyphicon-edit" />
      </button>
      <button
        className="deleteButton"
        onClick={[Function]}
        title="Remove this track">
        <span
          className="glyphicon glyphicon-trash" />
      </button>
    </li>
  </ul>
</div>
`;
